Fixes non-silicons being able to use machines without being adjacent

Irafas opened the pull request:
Checks to see if the user of a machine is both not a silicon and not standing next to it. If they are both the machine will not work.

:cl:
fix: non-silicons can no longer use machines without standing next to them.
/:cl:

fixes: #38916 
fixes: #38959

It wasn't just lathes, most machines were being effected.

Irafas committed:
Fixes non-silicons being able to use machines without being adjacent

-Checks to see if the user of a machine is both not a silicon and not standing next to it. If they are both the machine will not work.

ExcessiveUseOfCobblestone commented:
TK?

Irafas commented:
ah yeah forgot about that. Ill stick a check in. Thanks for reminding me.

Irafas committed:
TK users can use machines

-Added check to make sure that if the user has TK they can use machines remotely

Irafas committed:
only children of human type can have TK

-Only human types can actually have TK

Irafas committed:
Just a bit of added safety

Dorsisdwarf commented:
does this work with admin ai-interact?

ExcessiveUseOfCobblestone commented:
@Dorsisdwarf 

https://github.com/tgstation/tgstation/pull/38939/files#diff-64a44628db9b8e4a8f5e52467d26fb9dL215

ninjanomnom commented on code/game/machinery/_machinery.dm:
```diff
@@ -222,6 +222,10 @@ Class Procs:
 		return FALSE
 	else if(silicon && !(interaction_flags_machine & INTERACT_MACHINE_ALLOW_SILICON))
 		return FALSE
+	if(!silicon && !Adjacent(user))
```
you shouldn't need to check the silicon value 3 times in a row here, restructure the ifs a little

Irafas committed:
Restructured If's

-Restructered If's so that the var 'silicon' is checked less.

ninjanomnom commented on code/game/machinery/_machinery.dm:
```diff
@@ -218,10 +218,16 @@ Class Procs:
+	if(silicon)
+		if(!(interaction_flags_machine & INTERACT_MACHINE_ALLOW_SILICON))
+			return FALSE
+	else if(!silicon)
```
if this else is running you already know silicon is false

Irafas replied:
whoops. I meant to remove that before committing. Wanted to check if `!sillicon` would still evaluate as true if the var was null.

Irafas committed:
changed 'if else' to just 'else'

ninjanomnom commented on code/game/machinery/_machinery.dm:
```diff
@@ -218,10 +218,17 @@ Class Procs:
+		if(!Adjacent(user))
+			var/mob/living/carbon/H = user
+			if(istype(H) && H.has_dna() && !H.dna.check_mutation(TK))
+				return FALSE
```
If they're not adjacent but they have no dna they can interact?

Irafas replied:
I'm just looking to see if they have the TK power, as in that case they should be able to interact with machines without being adjacent. The `istype(H) && H.has_dna()` are just there to make sure the user has DNA to have checked in the first place. If the user doesn't have DNA they would definitely not have the TK power and so would be unable to interact with a machine remotely. 

So if they want to interact with a machine, without being a silicon, they would have to have both DNA and the TK power.

Would the `!H.dna.check_mutation(TK)` cause an error if it were to run without `H` having DNA to have checked?

ninjanomnom replied:
As it is currently, if they don't have the tk power because they don't have dna or aren't a carbon then this proc will allow interaction even though they're failing `Adjacent(user)`

Irafas replied:
Ah right. When I made this I assumed that mobs without DNA would never be able to interact with a machine anyway. So they would not even reach the proc.

That was probably not a great assumption to make. When I get back home I'll adjust it so mobs without DNA will be denied interaction as well. That would be safer.

Irafas committed:
Fixes noncarbon beings and those without DNA not being denied interaction

-Unless user is a carbon being that has DNA and the TK mutation they will be denied interaction.
